akka-http-core/src/main/scala/akka/http/engine/client/HostConnectionPoolSlot.scala

```diff
@@ -0,0 +1,239 @@
+               +-----------+                       +-------------+                +-----+------+                                    
+                                                                                        | SlotEvent                                                    
+                                                                                        | (to Conductor
+                                                                                        |  via slotEventMerge)
+                                                                                        v 
+   */
+  def apply(slotIx: Int, connectionFlow: Flow[HttpRequest, HttpResponse, Any],
+            remoteAddress: InetSocketAddress, // TODO: remove after #16168 is cleared
+            settings: HostConnectionPoolSettings)(implicit system: ActorSystem, fm: FlowMaterializer): Graph[Ports, Any] =
+    FlowGraph.partial() { implicit b ⇒
+      import FlowGraph.Implicits._
+
+      val slotProcessor = b.add {
+        Flow[RequestContext] andThenMat { () ⇒
+          val actor = system.actorOf(Props(new SlotProcessor(slotIx, connectionFlow, settings)), slotProcessorActorName.next())
+          (ActorProcessor[RequestContext, List[ProcessorOut]](actor), ())
+        }
+      }
+      val flatten = Flow[List[ProcessorOut]].mapConcat(identityFunc)
+      val split = b.add(new SlotEventSplit)
+
+      slotProcessor ~> flatten ~> split.in
+      Ports(slotProcessor.inlet, split.out0, split.out1)
+    }
+
+  import ActorSubscriberMessage._
+  import ActorPublisherMessage._
+
+  private class SlotProcessor(slotIx: Int, connectionFlow: Flow[HttpRequest, HttpResponse, Any],
+                              settings: HostConnectionPoolSettings)(implicit fm: FlowMaterializer)
+    extends ActorSubscriber with ActorPublisher[List[ProcessorOut]] with ActorLogging {
+
+    var exposedPublisher: akka.stream.impl.ActorPublisher[Any] = _
+    var inflightRequests = immutable.Queue.empty[RequestContext]
+    val runnableFlow = Source.actorPublisher[HttpRequest](Props(new FlowInportActor(self)))
+      .via(connectionFlow)
+      .toMat(Sink.actorSubscriber[HttpResponse](Props(new FlowOutportActor(self))))(_ -> _)
+
+    def requestStrategy = ZeroRequestStrategy
+    def receive = waitingExposedPublisher
+
+    def waitingExposedPublisher: Receive = {
+      case ExposedPublisher(publisher) ⇒
+        exposedPublisher = publisher
+        context.become(waitingForSubscribePending)
+      case other ⇒ throw new IllegalStateException(s"The first message must be `ExposedPublisher` but was [$other]")
+    }
+
+    def waitingForSubscribePending: Receive = {
+      case SubscribePending ⇒
+        exposedPublisher.takePendingSubscribers() foreach (s ⇒ self ! ActorPublisher.Internal.Subscribe(s))
+        context.become(unconnected)
+    }
+
+    val unconnected: Receive = {
+      case OnNext(rc: RequestContext) ⇒
+        assert(totalDemand > 0)
```

Comment:
Nope, this can also be the result of a RS spec violation of the upstream Publisher, in which case shutting down the ActorSystem without even printing a log seems a little harsh ;-) Seriously, using `assert` in Actors is (currently) not ideal.